### ISSUE TITLE
Templated Razor delegates?

### ISSUE THREAD

**Issue by naasking** [ltrzesniewski/RazorBlade#17]
Sorry for the series of issues, I'm prototyping some small projects with RazorBlade, RazorSlices and razor components in ways that I think I'll need to see where the pain points are. [Templated Razor delegates](https://learn.microsoft.com/en-us/aspnet/core/mvc/views/razor?view=aspnetcore-9.0#templated-razor-delegates) was the next thing I tried, since I want to create reusable markup fragments that can render child markup. If you have interest in covering this feature, here's a simple test case:
````html
@inherits RazorBlade.HtmlTemplate
@{
    Func<dynamic, object> span = @<span>hello!</span>;
}
<!DOCTYPE html>
<html lang="en">
<head><title></title></head>
<body>
</body>
</html>
````
Compilation fails with:
````
1>CSC : error RZ2000: The document type 'default' does not support the extension 'ITemplateTargetExtension'.
1>C:\repos\SqsRB\Templates2\Basic.cshtml(3,33,3,33): error CS1525: Invalid expression term ';'
````
The first line is the important one. RazorLight had a similar issue that was fixed in [this ticket](https://github.com/toddams/RazorLight/issues/236), if that's any help. The changes don't look too extensive.

Thanks again for all of your help!

**Comment by ltrzesniewski** [comment]
No worries 🙂 

I usually write a function in `@functions` for this kind of stuff, but I'll take a look at this.

**Comment by naasking** [comment]
I'm not sure you can use functions for this in a fully reusable fashion. Just as an example, suppose I have a styled button component whose contents should be up to the caller:

````html
@inherits RazorBlade.HtmlTemplate<Func<dynamic, object>>
<button class="flex items-center justify-center w-full px-4 py-2 text-sm font-medium leading-5 text-white text-gray-700 transition-colors duration-150 border border-gray-300 rounded-lg dark:text-gray-400 active:bg-transparent hover:border-gray-500 focus:border-gray-500 active:text-gray-500 focus:outline-none focus:shadow-outline-gray">
    @Model(default(dynamic))
</button>
````
Now I can construct buttons with simple or complicated content, like:

````html
<!-- razor strips <text> tags so this is plain text -->
@(new Button(@<text>some plain text</text>))

<!-- add an svg icon + text -->
@(new Button(@<text><svg class="w-4 h-4 mr-2" aria-hidden="true" viewBox="0 0 24 24" fill="currentColor">
                  <path d="M12 .297c-6.63 0-12 5.373-12 12 0 5.303 3.438 9.8 8.205 11.385.6.113.82-.258.82-.577 0-.285-.01-1.04-.015-2.04-3.338.724-4.042-1.61-4.042-1.61C4.422 18.07 3.633 17.7 3.633 17.7c-1.087-.744.084-.729.084-.729 1.205.084 1.838 1.236 1.838 1.236 1.07 1.835 2.809 1.305 3.495.998.108-.776.417-1.305.76-1.605-2.665-.3-5.466-1.332-5.466-5.93 0-1.31.465-2.38 1.235-3.22-.135-.303-.54-1.523.105-3.176 0 0 1.005-.322 3.3 1.23.96-.267 1.98-.399 3-.405 1.02.006 2.04.138 3 .405 2.28-1.552 3.285-1.23 3.285-1.23.645 1.653.24 2.873.12 3.176.765.84 1.23 1.91 1.23 3.22 0 4.61-2.805 5.625-5.475 5.92.42.36.81 1.096.81 2.22 0 1.606-.015 2.896-.015 3.286 0 .315.21.69.825.57C20.565 22.092 24 17.592 24 12.297c0-6.627-5.373-12-12-12"></path>
                </svg> Github</text>))
````
I don't think you can parameterize like this with named functions, and I don't think functions can be shared across templates. You can of course pass the content as an escaped string for this specific case, but manually escaping strings is less than ideal, and strictly less flexible since the template delegate lets you reference local parameters.

**Comment by ltrzesniewski** [comment]
I've implemented basic support for templated delegates in 8661d4b77dd5f488e77e86349696e83e162b6096, but I also had to add support for `PushWriter`/`PopWriter` and I'm not sure I'm satisfied with the resulting code, which became more complicated than necessary... Maybe I'll refactor it.

Anyway, it'll be shipped in v0.7.0

**Comment by naasking** [comment]
Awesome!
